Added custom callout example

ClayC-WeatherBug opened the pull request:
Also moved all annotations example views into an annotations subfolder

ferdicus commented on example/src/examples/Annotations/CustomCallout.tsx:
```diff
@@ -0,0 +1,93 @@
+
+const defaultCamera = {
+  centerCoordinate: [12.338, 45.4385],
+  zoomLevel: 17.4,
```
oddly specific :)

ClayC-WeatherBug replied:
Hah I copied this from one of our the examples 😁

ferdicus commented on example/src/examples/Annotations/CustomCallout.tsx:
```diff
@@ -0,0 +1,93 @@
+  message: String
+};
+
+const calloutContainerStyle:ViewStyle = {
```
nitpick: space after `:`

ClayC-WeatherBug replied:
Will do - can we add this to the prettier or eslint so that it complains about this and so that we can auto-fix it? If so I can include that in this PR.

ferdicus replied:
Hmmm, it should be working... when I have your branch open in my editor (vim in this case) and I save, this here:
```js
const calloutContainerStyle:ViewStyle = {
  backgroundColor: 'white',
  width: 60,
  height: 40,
  display: 'flex',
  justifyContent: 'center',
  alignItems: 'center'
};
```

turns into this

```js
const calloutContainerStyle: ViewStyle = {
  backgroundColor: 'white',
  width: 60,
  height: 40,
  display: 'flex',
  justifyContent: 'center',
  alignItems: 'center',
};

```

notice the space after `:` and a dangling `,` at the last object entry. 

I noticed though, that I missed to add a `devDependency` (`eslint-plugin-import`) to the `package.json`, 
maybe that caused an issue while linting in the background? 

Will add to `master` - would be cool if you could test it out on your branch :)

ClayC-WeatherBug replied:
Sounds great. I'll rebase from master, do a yarn install, and see if that fixes eslint not complaining. I'm using VSCode for this and use eslint/tslint/prettier on my other projects so it should work properly.

ferdicus commented on example/src/examples/Annotations/CustomCallout.tsx:
```diff
@@ -0,0 +1,93 @@
+  </View>;
+}
+
+const CustomCallout: FC<any> = (props) => {
```
Passing back the comment you left on my PR (which I shamefully haven't fixed yet 😓 ):
`any` props, no bueno 😭

ClayC-WeatherBug replied:
Hah - yeah, I'll try to figure it out. I copied your code as the base of this component to save some time. I'll figure out what props we are passing in so that I can create a type for our generic example props.

ferdicus commented on example/src/examples/Annotations/CustomCallout.tsx:
```diff
@@ -0,0 +1,93 @@
+}
+
+const CustomCallout: FC<any> = (props) => {
+  const [selectedFeature, setSelectedFeature] = useState<Feature<{ type: string; coordinates: number[]; }, any>>();
```
wow, I'm so used to always setting an initial state... I kinda forgot one can omit it :D

ferdicus commented on example/src/examples/Annotations/CustomCallout.tsx:
```diff
@@ -0,0 +1,93 @@
+const CustomCallout: FC<any> = (props) => {
+  const [selectedFeature, setSelectedFeature] = useState<Feature<{ type: string; coordinates: number[]; }, any>>();
+  const onPinPress = (e:any)=>{
+    if (e?.features?.length > 0) {
```
nice use of optional chaining, however why being explicit about the `length`?
wouldn't just `if (e?.features?.length) {` work as well?

ClayC-WeatherBug replied:
Weird I could have sworn that I replied to this. Yep, just doing e?.features?.length would definitely work as well because 0, undefined, and null are all falsey. I went with doing a length numerical check specifically because I think it makes it easier for readability. It makes it more like I'm checking if the length is greater than 0 rather than checking if length is a property or function on features. I'm totally fine with removing the > 0 if that is preferred though.

ferdicus replied:
I guess it's more a preference thing than anything else. 
I think it's ok to set back my personal preference for the sake of being more explicit here. 


all good 👍

ferdicus commented on example/src/examples/Annotations/CustomCallout.tsx:
```diff
@@ -0,0 +1,93 @@
+  const [selectedFeature, setSelectedFeature] = useState<Feature<{ type: string; coordinates: number[]; }, any>>();
+  const onPinPress = (e:any)=>{
+    if (e?.features?.length > 0) {
+      const feature = e?.features[0];
```
if ya wanna be fancy:
`const [feature] = e?.features;`

ClayC-WeatherBug replied:
Oh yeah good call. I tried to keep my syntax _mostly_ straightforward to hopefully make the example a bit easier for people less versed in JS to pick up. Do you prefer array destructuring in cases like this?

ferdicus replied:
if it's something simple like this - yeah I think I'd prefer that.
I think there is currently no rule enabled for this either in our config, so eslint wouldn't yell anyways. 

I think it's fine - as you said, for people who are less versed in JS, this would be easier to decipher than destructuring

ferdicus commented on example/src/examples/Annotations/CustomCallout.tsx:
```diff
@@ -0,0 +1,93 @@
+  return (
+    <Page
+      {...props}>
+      <MapboxGL.MapView style={sheet.matchParent}>
+        <MapboxGL.Camera defaultSettings={defaultCamera} />
+        <MapboxGL.ShapeSource
+                    id='mapPinsSource'
+                    shape={featureCollection}
+                    onPress={onPinPress}
+                >
+                    <MapboxGL.SymbolLayer
+                        id='mapPinsLayer'
+                        style={{ iconAllowOverlap: true,
+                            iconAnchor: 'bottom',
+                            iconSize: 1.0,
+                            iconImage: exampleIcon
+                        }}
+                    />
+                </MapboxGL.ShapeSource>
+                {selectedFeature && <MapboxGL.MarkerView id='selectedFeatureMarkerView'
+                    coordinate={selectedFeature.geometry.coordinates}>
+                    <CustomCalloutView
+                        message={selectedFeature?.properties?.message}
+                    />
+                </MapboxGL.MarkerView>}
+      </MapboxGL.MapView>
+    </Page>
+  );
```
mixing 2-4 spaces here - I'm curious, didn't eslint/ tslint complain?

ClayC-WeatherBug replied:
🤦  whoops and nope. VSCode + ESLint sometimes falls over on me and stops complaining about things. I moved some of the basic structure for this from a different project that uses 4 spaces and since it didn't complain I forgot to adjust it. Otherwise I would have auto-fixed it. I'll fix it in my PR update commit.

ferdicus reviewed (approved):
Thanks 👍 

Can you remove the `yarn.lock` file please :)

ferdicus commented:
@mfazekas, I think it's ok to add the `yarn.lock` file to the repo now. 
We can be a bit opinionated about that, what do you think?

ClayC-WeatherBug commented:
> @mfazekas, I think it's ok to add the `yarn.lock` file to the repo now.
> We can be a bit opinionated about that, what do you think?

What was the reason for excluding yarn.lock updates before? If we include them in a PR then I believe the yarn.lock should only update in future PRs whenever we modify our package versions.

mfazekas commented:
> > @mfazekas, I think it's ok to add the `yarn.lock` file to the repo now.
> > We can be a bit opinionated about that, what do you think?
> 
> What was the reason for excluding yarn.lock updates before? If we include them in a PR then I believe the yarn.lock should only update in future PRs whenever we modify our package versions.

There are arguments against checking in `lock files`, esp since we're a library.
See: https://dev.to/gajus/stop-using-package-lock-json-or-yarn-lock-3ddi

ferdicus commented:
> > > @mfazekas, I think it's ok to add the `yarn.lock` file to the repo now.
> > > We can be a bit opinionated about that, what do you think?
> > 
> > 
> > What was the reason for excluding yarn.lock updates before? If we include them in a PR then I believe the yarn.lock should only update in future PRs whenever we modify our package versions.
> 
> There are arguments against checking in `lock files`, esp since we're a library.
> See: https://dev.to/gajus/stop-using-package-lock-json-or-yarn-lock-3ddi

I propose the following then:
* ignore lock files in root
* add lock file for `/example`

Objections?

mfazekas commented:
> I propose the following then:
> 
>     * ignore lock files in root
> 
>     * add lock file for `/example`
> 
> 
> Objections?

Since our CI is working with the example app, i'd vote to ignore the example in lockfile too.

ferdicus commented:
> > I propose the following then:
> > ```
> > * ignore lock files in root
> > 
> > * add lock file for `/example`
> > ```
> > 
> > 
> > Objections?
> 
> Since our CI is working with the example app, i'd vote to ignore the example in lockfile too.

👍

ClayC-WeatherBug commented:
@ferdicus @mfazekas I updated the PR based on the PR comments. These were the changes:
- Added yarn.lock to the gitignore in examples and at the root of the project
- Forgot the examples yarn.lock file
- Fixed spacing where it was messed up in my example
- Added prop types for the CustomCallout component

Here is a screenshot of the example also. I can pretty it up if you think that would be good, but I figured it would be better to keep the code as simple as possible since it's an example in order to avoid confusion. Please let me know either of you have any other things you would like for me to address before merging.

![image](https://user-images.githubusercontent.com/65256865/93098673-232d4f80-f675-11ea-9fc8-895f04080fb1.png)

ferdicus commented on example/src/examples/Annotations/CustomCallout.tsx:
```diff
@@ -0,0 +1,97 @@
+import React, {FC, useState, useEffect} from 'react';
```
`useEffect` is not being used, right?

ClayC-WeatherBug replied:
Whoops - I'll double-check. I may have added it initially for something else and forgot to remove. Thanks

ferdicus commented on example/src/examples/Annotations/CustomCallout.tsx:
```diff
@@ -0,0 +1,97 @@
+  ]
+};
+
+type CustomCalloutViewProps = {
+  message: String
+};
```
👍

ferdicus commented on example/src/examples/Annotations/CustomCallout.tsx:
```diff
@@ -0,0 +1,97 @@
+
+const CustomCallout: FC<CustomCalloutProps> = (props) => {
+  const [selectedFeature, setSelectedFeature] = useState<Feature<{ type: string; coordinates: number[]; }, any>>();
+  const onPinPress = (e:any)=>{
```
when you have the es-/ts-lint rules running, it will complain here, that this function does not provide a return type. 
`:void` should do it.

I enabled that rule per default - let me know if you think of it as non-sense or if you agree :)

ClayC-WeatherBug replied:
I'm fine with having to do explicit return types here. I'll update 👍

ferdicus commented on example/src/examples/Annotations/CustomCallout.tsx:
```diff
@@ -0,0 +1,97 @@
+  message: String
+};
+
+const calloutContainerStyle: ViewStyle = {
+  backgroundColor: 'white',
+  width: 60,
+  height: 40,
+  display: 'flex',
+  justifyContent: 'center',
+  alignItems: 'center'
+};
```
I like this 💯 

Also, there is (of course) another lint rule, that'll complain about inline styles,   
which there are some in this file.

Suggestion:

```js
const styles = {
  calloutContainer: {
    backgroundColor: 'white',
    width: 60,
    height: 40,
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center'
  },
  calloutText: {
    color: 'black',
    fontSize: 16
  }
 // all other styles
};
```

what do you think?

ClayC-WeatherBug replied:
Yep, I can make that change. I think that works.

ferdicus reviewed (approved):
> @ferdicus @mfazekas I updated the PR based on the PR comments. These were the changes:
> 
>     * Added yarn.lock to the gitignore in examples and at the root of the project
> 
>     * Forgot the examples yarn.lock file
> 
>     * Fixed spacing where it was messed up in my example
> 
>     * Added prop types for the CustomCallout component
> 
> 
> Here is a screenshot of the example also. I can pretty it up if you think that would be good, but I figured it would be better to keep the code as simple as possible since it's an example in order to avoid confusion. Please let me know either of you have any other things you would like for me to address before merging.
> 
> IMAGE

Looks great 👍 

Found some minor things to nitpick :D

ferdicus commented:
hey @ClayC-WeatherBug, feel free to Squash n Merge once you committed the changes you described. 

Cheers 🍻

ClayC-WeatherBug commented:
> hey @ClayC-WeatherBug, feel free to Squash n Merge once you committed the changes you described.
> 
> Cheers 🍻

Thanks! Sorry for the delay on merging this. I got wrapped up on some other things this week and haven't had a chance to update the PR. I'll try to do it sometime early next week if at all possible.

ferdicus commented:
@ClayC-WeatherBug, hey, sorry to be such a pain in the butt... can we get this done and merged? 
I feel like there isn't much left to be done.

Thanks in advance 🙇

ClayC-WeatherBug commented:
> @ClayC-WeatherBug, hey, sorry to be such a pain in the butt... can we get this done and merged?
> I feel like there isn't much left to be done.
> 
> Thanks in advance 🙇

Yeah - truly sorry that this fell off my radar. I got real busy with some work things and hadn't had a chance to circle back around to this. I'll see if I can knock it out today or early next week. Thanks for the reminder 👍

ClayC-WeatherBug commented:
PR comments addressed - merging. Thanks for the review!
